### DISCUSSION TITLE
Consider using Token Types

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#905] [Ideas]
Type safety for ERC-20 and ERC-721 interactions:

https://github.com/jtriley-eth/token-types

CC @sablier-labs/solidity.

**Comment by PaulRBerg** [comment]
did you have a look at this, @sablier-labs/solidity?

**Reply by andreivladbrg** [reply]
Actually not, I think I've missed this discussion, will do, but a low priority atm

**Reply by smol-ninja** [reply]
Not yet, but will do a review soon.

**Reply by PaulRBerg** [reply]
cheers guys

**Comment by smol-ninja** [comment]
### Good stuffs

- The implementations in token-types contracts are memory safe, so using them give us a slight gas advantage over OZ but its only ~1% ([tx1](https://sepolia.etherscan.io/tx/0x7b6853df8b2e05c0df014c57f78ea9a7270e5d0a589ed0178b178f816e70d42c) vs [tx2](https://sepolia.etherscan.io/tx/0x40ea540ac1d2d2d63df170c47765c92936b71b8123af1bfb403ad57a2a5f5a26))
- In case of ERC20, the implementation of `transfer` and `transferFrom` are similar to `safeTransfer` and `safeTransferFrom` from `SafeERC20`, so we won't need the `SafeERC20` library anymore.

### Bad stuffs

- In token-types ERC20, there is no implementation akin to `forceApprove` so thats something we will have to handle or use `SafeERC20` alongside it.
- But `SafeERC20` library cannot be used out of the box because `SafeERC20` takes `IERC20` as the first input.  So, in order to use it, we will have to pass `IERC20(address(token))` as the first parameter of `forceApprove`.
- Only 1 release so far with no new PR or issue.

### Note
- To differentiate it from OZ's ERC20, this should be imported as `IERC20` or some other alias.
- Minimum `0.8.25` is required which is fine by us.

### My take

- I don't see much advantage in using Token Type over traditional interfaces.

**Reply by PaulRBerg** [reply]
Thanks for taking a look and for the clear breakdown.

> Only 1 release so far with no new PR or issue.

Strong argument against using it.

The security risks are not worth it.
